domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java

```diff
@@ -580,79 +594,73 @@ public static String pojo2json(Object entity) throws Exception {
    * both the connection and the transaction remain open to let the caller
    * decide what to do.
    *
-   * @see #endTx(Object, Handler)
-   * @see #rollbackTx(Future, Handler)
    * @param done - the result is the current connection
    */
-  //@Timer
   public void startTx(Handler<AsyncResult<SQLConnection>> done) {
-    client.getConnection(res -> {
-      if (res.succeeded()) {
-        SQLConnection connection = res.result();
-        try {
-          connection.setAutoCommit(false, res1 -> {
-            if (res1.failed()) {
-              connection.close();
-              done.handle(Future.failedFuture(res1.cause()));
-            } else {
-              done.handle(Future.succeededFuture(connection));
-            }
-          });
-        } catch (Exception e) {
-          log.error(e.getMessage(), e);
-          if (connection != null) {
-            connection.close();
-          }
-          done.handle(Future.failedFuture(e));
-        }
-      }
-      else{
+    getConnection(res -> {
+      if (res.failed()) {
         log.error(res.cause().getMessage(), res.cause());
         done.handle(Future.failedFuture(res.cause()));
+        return;
+      }
+      try {
+        SQLConnection pgTransaction = new SQLConnection(res.result(), res.result().begin());
+        done.handle(Future.succeededFuture(pgTransaction));
+      } catch (Exception e) {
+        log.error(e.getMessage(), e);
+        done.handle(Future.failedFuture(e.getCause()));
       }
     });
   }
 
+  static void finalizeTx(AsyncResult<Void> txResult, SqlConnection conn, Handler<AsyncResult<Void>> done ) {
+    if (conn != null) {
+      conn.close();
+    }
+    if (txResult.failed() && !"Transaction already completed".equals(txResult.cause().getMessage())) {
```

Comment:
It would be nice not to have to compare to the transaction exception message. But I do not see a way to get the integer status.

Reply:
Yeah.. Not pretty.. Reverse mapping those message are really problematic.